File: C_plus_plus/Best_Time_to_Buy_and_Sell_Stock_II.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int start,end;
        int max = 0;
        if(prices.size()==0) return 0;
        start = end = prices[0];
        for(int i = 1; i < prices.size(); ++i){
            if(prices[i]>=prices[i-1]){
                end = prices[i];
            }
            else{
                max += end - start;
                start = prices[i];
                end = prices[i];
            }
        }
        max += end - start;
        return max;
    }
};

File: java/Best_Time_to_Buy_and_Sell_Stock_II.java
// One pass solution
// Just compare the current with the previous one, if there is a profit , we buy and sell at the two days respectively.
// Time is O(n), space is O(1).

public class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        if (n == 0) return 0;
        int premin = prices[0], profit = 0;
        for (int i = 1; i < n; ++i) {
            if (prices[i] > premin)
                profit += prices[i] - premin;
            premin = prices[i];
        }
        return profit;
    }
}

// We can simplize it like this:

public class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length == 0) return 0;
        int max = 0;
        for (int i = 1; i < prices.length; ++i) {
            if (prices[i] >= prices[i-1])
                max += prices[i] - prices[i-1];
        }
        return max;
    }
}
